include/envoy/router/router.h

```diff
@@ -416,7 +430,7 @@ class Decorator {
 typedef std::unique_ptr<const Decorator> DecoratorConstPtr;
 
 /**
- * An interface that holds a RedirectEntry or a RouteEntry for a request.
```

Comment:
I'd prefer this replace RedirectEntry rather than be a slightly duplicate thing.
newPath could be optional, and expected for redirect responses.   I'm neutral if it should be done here or in a follow-up but maybe add a TODO?

Reply:
Thanks for the suggestion; combining RedirectEntry and DirectResponseEntry sounds good to me. The combined code will have to deal with some incompatibilities between the two -- e.g., under the current data plane API, redirects aren't allowed to send the `response_headers_to_add`, whereas direct responses are required to -- but it will probably still be cleaner than the current diff.

I'll merge the implementations post an update to this PR later today.

Reply:
cool.  hopefully if we do some sanity checks with an IsRedirectReturnCode() in the right places we'll guard against bad configs/code

Reply:
I don't see any reasons redirects cannot also populate response_headers_to_add. That seems pretty useful to me. Thoughts?

Reply:
Adding response_headers_to_add to redirects would change the behavior of existing configs. If we were designing the config schema from scratch, I’d advocate for sending the headers with redirects. But at this point it would impact backward compatibility. Unless we added a “send_headers_with_redirects” flag that defaults to false, I suppose :(

Reply:
Yeah, fair enough. I kind of think we could "just do it" and document it in this case as a change. But that's me. Fine to block/TODO for now if you think that would be best.